File: pulsar/java/src/main/java/com/examples/hello/pulsar/consumers/ConsumerFactory.java
package com.examples.hello.pulsar.consumers;

import com.examples.hello.pulsar.weather.conditions.Condition;
import com.examples.hello.pulsar.consumers.batch.BatchConsumerRunnable;
import com.examples.hello.pulsar.consumers.batch.BatchConsumerRunner;
import com.examples.hello.pulsar.consumers.batch.ConditionBatchConsumer;
import com.examples.hello.pulsar.services.PulsarFactory;
import io.micronaut.context.annotation.Factory;
import io.micronaut.context.annotation.Value;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import lombok.extern.slf4j.Slf4j;
import lombok.val;

@Factory
@Slf4j
public class ConsumerFactory {
  @Inject
  private PulsarFactory factory;

  @Value("${pulsar.consumer.name}")
  String name;
  @Value("${pulsar.consumer.topic}")
  String topic;

  @Singleton
  public ConsumerRunnable consumer() {
    val consumer = factory.consumerOf(name, topic, Condition.class);
    return new ConsumerRunner<>(consumer, sink());
  }

  @Singleton
  public BatchConsumerRunnable batchConsumer() {
    val consumer = factory.consumerOf(name, topic, Condition.class);
    return new BatchConsumerRunner<>(consumer, batchSink());
  }

  private ConditionConsumer sink() {
    return msg -> log.debug("Received condition={}", msg.getValue());
  }

  private ConditionBatchConsumer batchSink() {
    return msg -> log.debug("Received conditions batch (size={})", msg.size());
  }
}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/consumers/ConsumerFactory.kt
package com.example.hello.pulsar.consumers

import com.example.hello.pulsar.Condition
import com.example.hello.pulsar.consumers.batch.BatchConsumerRunnable
import com.example.hello.pulsar.consumers.batch.BatchConsumerRunner
import com.example.hello.pulsar.services.PulsarFactory
import com.example.hello.pulsar.services.PulsarFactory.Companion.shared
import io.micronaut.context.annotation.Factory
import io.micronaut.context.annotation.Value
import jakarta.inject.Inject
import jakarta.inject.Singleton
import org.apache.pulsar.client.api.Message
import org.slf4j.LoggerFactory

@Factory
@Singleton
class ConsumerFactory(
  @Inject private val factory: PulsarFactory,
  @Value("\${pulsar.consumer.name}") val name: String,
  @Value("\${pulsar.consumer.topic}") val topic: String
) {
  @Singleton
  fun consumer(): ConsumerRunnable {
    val consumer = factory.weatherConsumer(topic).shared(name).consumerName(name).subscribe()
    return ConsumerRunner(consumer, this::sink)
  }

  @Singleton
  fun batchConsumer(): BatchConsumerRunnable {
    val consumer = factory.weatherConsumer(topic).shared(name).consumerName(name).subscribe()
    return BatchConsumerRunner(consumer, this::batchSink)
  }

  private fun sink(msg: Message<Condition>) {
    log.debug("Received condition={}", msg.value)
  }

  private fun batchSink(msg: List<Message<Condition>>) {
    log.debug("Received conditions batch (size={})", msg.size)
  }

  companion object {
    private val log = LoggerFactory.getLogger(ConsumerFactory::class.java)
  }
}
